measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+                raise ValueError(
+                    "error_cov[i] must have the same length as total_measure_idx. Expect number of columns:"
+                    + str(self.sens_info.total_measure_idx)
+                )
+
+        elif error_option == CovarianceStructure.time_measure_correlation:
+            # check row number
+            if len(error_cov) != self.total_num_time:
+                raise ValueError(
+                    "error_cov must have the shape total_num_time*total_num_time. Expect number of rows:"
+                    + str(self.total_num_time)
+                )
+            # check column number
+            if len(error_cov[0]) != self.total_num_time:
+                raise ValueError(
+                    "error_cov must have the shape total_num_time*total_num_time. Expect number of columns:"
+                    + str(self.total_num_time)
+                )
+
+    def _dynamic_flatten(self, jac):
+        """Update dynamic flattened matrix index.
+        Arguments
+        ---------
+        :param jac: Jacobian information for main class use, a 2D array of shape [N_total_meausrements * Np]
+
+        Returns
+        -------
+        dynamic_flatten matrix: flatten dynamic-cost measurements, not flatten static-costs, [s1, d1|t1, ..., d1|tN, s2]
+        Flatten matrix: flatten dynamic-cost and static-cost measuremenets
+        """
+
+        ### dynamic_flatten: to be decision matrix
+        jac_dynamic_flatten = []
+        # position index in jac_dynamic_flatten where each measurement starts
+        self.head_pos_dynamic_flatten = {}
+        # all static measurements index after dynamic_flattening
+        self.static_idx_dynamic_flatten = []
+        self.dynamic_idx_dynamic_flatten = []
+
+        ### flatten: flatten all measurement all costs
+        jac_flatten = []
+        # position index in jac_flatten where each measurement starts
+        self.head_pos_flatten = {}
+        # all static measurements index after flatten
+        self.static_idx_flatten = []
+        # all dynamic measurements index after flatten
+        self.dynamic_idx_flatten = []
+
+        # map dynamic index to flatten index
+        # key: dynamic index, value: corresponding indexes in flatten matrix. For static, it's a list. For dynamic, it's a index value
+        self.dynamic_to_flatten = {}
+
+        # counter for dynamic_flatten
+        count1 = 0
+        # counter for flatten
+        count2 = 0
+        # loop over total measurement index
+        for i in range(self.sens_info.total_measure_idx):
+            if (
+                i in self.static_measurement_idx
+            ):  # static measurements are not flattened for dynamic flatten
+                if self.precompute_print_level == 3:
+                    print("Static-cost measurement idx: ", i)
+                # dynamic_flatten for SCM
+                jacobian_static = []
+                # locate head row of the sensitivity for this measurement
```

Comment:
Instead of "head" I think you mean "first"

Reply:
Please confirm and update variable names to be more clear.

Reply:
yes. updated